hr_attendance_missing_days/models/hr_employee.py

```diff
@@ -0,0 +1,72 @@
+# © 2023 initOS GmbH
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from datetime import datetime, time
+
+import pytz
+
+from odoo import models
+
+_logger = logging.getLogger(__name__)
+
+
+def ensure_tz(dt, tz=None):
+    if not dt.tzinfo:
+        dt = pytz.utc.localize(dt)
+    return dt.astimezone(tz) if tz else dt
+
+
+class Employee(models.Model):
+    _inherit = "hr.employee"
+
+    def create_missing_attendances(self, date_from=None, date_to=None):
+        for emp in self.search([]):
+            emp._create_missing_attendances(date_from, date_to)
+
+    def _create_missing_attendances(self, date_from=None, date_to=None):
+        self.ensure_one()
+
+        reason = self.env.company.sudo().attendance_missing_days_reason
+        if not reason:
+            return
+
+        if not date_from:
+            date_from = datetime.combine(
+                self.env.company.sudo().overtime_start_date, time.min
+            )
+
+        if not date_to:
+            date_to = datetime.now()
+
+        date_from, date_to = map(ensure_tz, (date_from, date_to))
+
+        intervals = self.resource_calendar_id._work_intervals_batch(date_from, date_to)
+        work_dates = {}
+        for start, _stop, _attendance in sorted(intervals[False]):
+            start_date = start.date()
+            if start_date not in work_dates:
+                work_dates[start_date] = ensure_tz(start, pytz.utc).replace(tzinfo=None)
+
+        domain = [
+            ("check_in", ">=", date_from.replace(tzinfo=None)),
+            ("check_in", "<=", date_to.replace(tzinfo=None)),
```

Comment:
this gives you wrong dates before 01:00/02:00 in timezone Europe/Amsterdam, in the US or more so NZ the error will be bigger.

```suggestion
            ("check_in", "<=", date_to.astimezone(pytz.utc).replace(tzinfo=None)),
```

same for date_from

Reply:
I found the problem. The problem was when the cron runs in the mid of a working day and had nothing to do with timezones. E.g. working day goes from 8:00 - 17:00 UTC. The cron runs 11:30 but the employee logs in at 12:00. The employee is present (only late) but the function created a 0 attendance. Fixed because the job filters out working days where the corresponding days aren't over yet.

Reply:
should the function maybe better work on dates (instead of datetime) anyways? Then it's clear which intervals are meant, and if you then create datetime objects as

```python
datetime_from = timezone(employee.tz).localize(
    datetime.combine(date_from, time.min)
).astimezone(pytz.utc).replace(tzinfo=None)

datetime_to = timezone(employee.tz).localize(
    datetime.combine(date_to, time.max)
).astimezone(pytz.utc).replace(tzinfo=None)
```

it's guaranteed you never miss an attendance from that date from the point of view of the employee's timezone.

This will also make it easier for fellow programmers to use the function, because they don't have to consider any timezone stuff.

With this all my other comments would be covered too

Reply:
@fkantelberg if you don't have time to implement this but generally agree with the proposal, I'll be happy to PR this to you PR

Reply:
I reworked it to use dates as argument and